Source/Particles/WarpXParticleContainer.cpp

```diff
@@ -646,25 +646,62 @@ std::array<Real, 3> WarpXParticleContainer::meanParticleVelocity(bool local) {
     amrex::Real inv_clight_sq = 1.0/PhysConst::c/PhysConst::c;
 
     const int nLevels = finestLevel();
-    for (int lev = 0; lev <= nLevels; ++lev) {
 
+#ifdef AMREX_USE_GPU
+    if (Gpu::inLaunchRegion())
+    {
+        ReduceOps<ReduceOpSum, ReduceOpSum, ReduceOpSum> reduce_op;
+        ReduceData<Real, Real, Real> reduce_data(reduce_op);
+        using ReduceTuple = typename decltype(reduce_data)::Type;
+        for (int lev = 0; lev <= nLevels; ++lev) {
```

Comment:
Could you explain a little bit what this is doing?

Reply:
This is how we can do reductions on the GPU - in particular, using the `ReduceTuple`, we can compute the total `vx`, `vy`, and `vz` in one kernel launch.

Reply:
Nice feature! Thanks for the explanation.